lib/core/error.js

```diff
@@ -153,40 +153,71 @@ function isRetryableError(error) {
   );
 }
 
-const SDAM_UNRECOVERABLE_ERROR_CODES = new Set([
+const SDAM_RECOVERING_CODES = new Set([
   91, // ShutdownInProgress
   189, // PrimarySteppedDown
-  10107, // NotMaster
   11600, // InterruptedAtShutdown
   11602, // InterruptedDueToReplStateChange
-  13435, // NotMasterNoSlaveOk
   13436 // NotMasterOrSecondary
 ]);
+
+const SDAM_NOTMASTER_CODES = new Set([
+  10107, // NotMaster
+  13435 // NotMasterNoSlaveOk
+]);
+
+const SDAM_NODE_SHUTTING_DOWN_ERROR_CODES = new Set([
+  11600, // InterruptedAtShutdown
+  91 // ShutdownInProgress
+]);
+
+function isRecoveringError(err) {
+  if (err.code && SDAM_RECOVERING_CODES.has(err.code)) {
+    return true;
+  }
+
+  return err.message.match(/not master or secondary/) || err.message.match(/node is recovering/);
+}
+
+function isNotMasterError(err) {
+  if (err.code && SDAM_NOTMASTER_CODES.has(err.code)) {
+    return true;
+  }
+
+  if (isRecoveringError(err)) {
+    return false;
+  }
+
+  return err.message.match(/not master/);
+}
+
+function isNodeShuttingDownError(err) {
+  return SDAM_NODE_SHUTTING_DOWN_ERROR_CODES.has(err.code);
```

Comment:
Do we need to check `if (err.code)` here?

Reply:
:+1: